### ISSUE TITLE
AutoPublishAlias with a !Ref value marked as false positive

### ISSUE THREAD

**Issue by milancermak** [aws-cloudformation/cfn-python-lint#582]
*cfn-lint version: 0.11.1*

I'm running into an issue with `AutoPublishAlias`. When using a `!Ref` as its value, I get the following error message:
```
E0001 Resource with id [SkillFunction] is invalid. 'AutoPublishAlias' must be a string or a Ref to a template parameter
foo.yml:1:1
```

Here's the minimal template to reproduce the issue:
```yaml
---

AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Parameters:
  Stage:
    Description: Environment stage (deployment phase)
    Type: String
    AllowedValues:
      - beta
      - prod

Resources:
  SkillFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: '.'
      Handler: main.handler
      Runtime: python3.7
      Timeout: 30
      MemorySize: 128
      AutoPublishAlias: !Ref Stage
```

The error message is actually very helpful (👏 for that), but as you can see, I am using a `!Ref` to a template parameter, so this should not cause a linting error.

**Comment by kddejong** [comment]
Taking a look at it.  Looks like we may be getting this error back from SAM when trying to transform the template.  So need to dig a little deeper.

**Comment by kddejong** [comment]
https://github.com/awslabs/serverless-application-model/blob/0425efc396b659c8d0dea5b63c3287cb915d6b34/samtranslator/model/sam_resources.py#L144

**Comment by kddejong** [comment]
This looks like an issue specifically with AutoPublishAlias and how SAM is designed. https://github.com/awslabs/serverless-application-model/issues/220

**Comment by milancermak** [comment]
First of, thank you for jumping on the issue so promptly.

Should I open a new issue in the SAM repo then? Not sure how you folks are organized.

**Comment by kddejong** [comment]
Sorry for the delay I think I may be out of my element on this one.  I get the same error from the CLI.  From what I can tell this shouldn't be happening but I'm a little bit lost on why.  I think we should create an incident in https://github.com/awslabs/serverless-application-model and see if they can help

```
2019-01-16 20:37:27 Found credentials in shared credentials file: ~/.aws/credentials
Template provided at 'test.yaml' was invalid SAM Template.
Error: [InvalidResourceException('SkillFunction', "'AutoPublishAlias' must be a string or a Ref to a template parameter")] ('SkillFunction', "'AutoPublishAlias' must be a string or a Ref to a template parameter")
```

**Comment by kddejong** [comment]
Yea sorry I didn't explain that well.  We can't lint a pre-transformed template because there isn't a CloudFormation spec for it.  We pass it to the sam translator and then lint the results.  In this case since they give us back an error we spit it back to you and don't lint the file.

**Comment by milancermak** [comment]
Alright, got it. Closing the issue then as it's not a cfn-lint error really.

**Comment by jlhood** [comment]
Hey, James from the SAM repo here. We were reviewing https://github.com/awslabs/serverless-application-model/issues/778 and we think this issue should be reopened (I don't seem to have permissions to do it). From the other issue:

> The issue here is that SAM actually needs to resolve the value of certain parameters in order to generate the logical ids of other CloudFormation resources during the transform. AutoPublishAlias is one example where this is true, which is why SAM is failing here. Since these parameters are required when actually deploying the template, it works fine when deploying direct to CloudFormation. However, it fails on validate, because CFN lint (and SAM CLI) are manually calling the SAM transform function without any parameter values in order to do validation.
> 
> The fix for this is either for the SAM translator library to offer a validate() function that accounts for issues like this under the covers, or for CFN lint to call transform with dummy parameter values.

Since SAM doesn't currently offer a `validate()` library function, the alternative fix is for CFN lint to generate dummy parameters when calling `translate()`.

https://github.com/awslabs/cfn-python-lint/blob/f5c3504f593d4311388a047fa4b6f18f792c919a/src/cfnlint/transform.py#L114

**Comment by jlhood** [comment]
@kddejong Can we get this reopened per my comment above?

**Comment by kddejong** [comment]
Trying to figure out what to change here.  https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/transform.py#L70-L84.  The trick here being I should also Remove the Stage Parameter.

```
if isinstance(resource_dict.get('AutoPublishAlias'), dict):
   resource_dict['AutoPublishAlias'] = 'Alias'
```

results in
```
W2001 Parameter Stage not used.
test.yaml:7:3
```

Even if I provide Parameter values the results are the same.  I have a few options here.  I can validate that the parameter isn't being used any where else and then remove it.  It will take me a little longer to get that written.

@jlhood I wonder if translate should remove the parameter if I provide the value for it... maybe you keep it that way for visibility?

**Comment by jlhood** [comment]
Hi @kddejong. Thanks for looking into this! Sorry for the delayed response. I was on vacation. 😊

We don't think the right fix is to modify the template to remove that property as you suggested. The key issue is the SAM translator can error if required template parameters (no Default value specified in template) do not have values. During actual CloudFormation deployment, CloudFormation fails before calling the SAM translator if the user does not supply required parameters which is why this error is only showing up on cfn-lint validation.

We think the correct fix is to supply dummy values for any required template parameters. Currently, cfn-lint is always passing an empty dict of parameter values:

https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/transform.py#L107

The fix would be to generate dummy parameter values for any required template parameters and pass those in to the SAM `translate()` function.

**Comment by kddejong** [comment]
Sounds fair.  I can work on it our side.  Trick from a lint-ing perspective is translate removes a Ref to a parameter and we will need to remove that parameter (if its no longer used) so we don't flag false positives.

**Comment by kddejong** [comment]
Sorry, lost sight of this.  I have a fix in for this now.

**Comment by cmmeyer** [comment]
@jlhood -- What other properties would require us to fill in a fake value for a !Ref? Is it a subset or "all"?

**Comment by jlhood** [comment]
@cmmeyer Took a look at the PR. The fix shouldn't be specific to `AutoPublishAlias`. It should fill in dummy values for any template parameters that do not specify a default value before calling SAM's translate function.

**Comment by cmmeyer** [comment]
@jlhood -- what other properties will fail?  It sounds like SAM validation is starting to be more deployment aware? We really are trying to stay as focused on the template as possible.

**Comment by jlhood** [comment]
@cmmeyer The key issue is sometimes SAM has to (partially) resolve intrinsic functions at transform time. When we see `!Ref` of a template parameter, SAM can resolve it. During an actual CFN deploy, this always works, because all parameters are guaranteed to have values (either default or passed in by the user since CFN enforces required parameters before the SAM transform is ever called).

However, in the cfn-lint scenario, this is not true which is what's causing this error. I think the ideal fix is for cfn-lint to populate dummy parameter values for template parameters that don't have a default value.
